### PR TITLE
Add Ground Spike Antenna

### PR DESCRIPTION
**Pull request by TheMagnetar**
**When merged this pull request will:**
- Adds antenna ground spike to acre2
- [x] Add item icons
- [x] Add item previews
- [x] Adjust mass
- [x] Hide connector actions
- [x] Better placement
- [x] Testing, testing and more testing.

### CODE REVIEW

File: addons/sys_core/fnc_hasItem.sqf [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+ * Example:
+ * [player, "acre2_vhf30108"] call acre_sys_core_fnc_hasItem
+ *
+ * Public: No
+ */
+#include "script_component.hpp"
+
+params [["_unit", objNull], ["_item", ""]];
+
+_item in (items _unit)
```
Comment (jonpas) [inline addons/sys_core/fnc_hasItem.sqf]: This is pretty slow, might be worth looking at https://github.com/acemod/ACE3/pull/6350 (first point of caching items and using `getItemCargo`).

File: addons/sys_gsa/CfgWeapons.hpp [hunk 2]
```diff
@@ -0,0 +1,47 @@
+#include "script_component.hpp"
+
+class CfgWeapons {
+    class CBA_MiscItem;
+    class CBA_MiscItem_ItemInfo;
+
+    class acre2_vhf30108: CBA_MiscItem {
```
Comment (jonpas) [inline addons/sys_gsa/CfgWeapons.hpp]: Why are those lower-case? We use `ACRE_` for radios and all lower-case, should probably make this the same. (Also easier to move to ACRE3! :P)
```suggestion
    class ACRE_VHF30108: CBA_MiscItem {
```
Same for others classes.

File: addons/sys_gsa/XEH_postInit.sqf [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    if (_connectedGsa select 0) exitWith {
+        [QGVAR(notifyPlayer), [localize LSTRING(alreadyConnected)], _player] call CBA_fnc_targetEvent;
+    };
+
+    // Force attach the ground spike antenna
+    [_radioId, 0, _componentName, [], true] call EFUNC(sys_components,attachSimpleComponent);
+
+    _gsa setVariable [QGVAR(connectedRadio), _radioId, true];
+    [_radioId, "setState", ["externalAntennaConnected", [true, _gsa]]] call EFUNC(sys_data,dataEvent);
+    systemChat format ["connected"];
```
Comment (jonpas) [inline addons/sys_gsa/XEH_postInit.sqf]: Delete.

File: addons/sys_gsa/XEH_postInit.sqf [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+    [QGVAR(notifyPlayer), [localize LSTRING(connected)], _player] call CBA_fnc_targetEvent;
+
+    call compile format [QUOTE(GVAR(%1) = [ARR_3(DFUNC(externalAntennaPFH), 1.0, [ARR_2(_gsa, _radioId)])] call CBA_fnc_addPerFrameHandler;), _radioId];
```
Comment (jonpas) [inline addons/sys_gsa/XEH_postInit.sqf]: ???

File: addons/sys_gsa/config.cpp [hunk 5]
```diff
@@ -0,0 +1,19 @@
+#include "script_component.hpp"
+
+class CfgPatches {
+    class ADDON {
+        name = COMPONENT_NAME;
+        units[] = {};
+        weapons[] = {};
+        requiredVersion = REQUIRED_VERSION;
+        requiredAddons[] = {"acre_main", "acre_sys_core"};
```
Comment (jonpas) [inline addons/sys_gsa/config.cpp:9]: `acre_sys_core` requires `acre_main` already. I don't know why we do that in other places either though.

File: addons/sys_gsa/data/models/acre_antennabag.rvmat [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+emmisive[] = {0, 0, 0, 0};
+specular[] = {0, 0, 0, 1};
+specularPower = 0;
+PixelShaderID = "NormalMap";
+VertexShaderID = "NormalMap";
+class Stage1 {
+    texture = "idi\acre\addons\sys_gsa\data\model\acre_antennaBag_nohq.paa";
+    uvSource = "tex";
+    class uvTransform {
+        aside[]={1, 0, 0};
```
Comment (jonpas) [inline addons/sys_gsa/data/models/acre_antennabag.rvmat]: Formatting.

File: addons/sys_gsa/fnc_connectChildrenActions.sqf [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+ *
+ * Example:
+ * [vehicle acre_player] call acre_sys_gsa_fnc_connectChildrenActions
+ *
+ * Public: No
+ */
+
+params ["_unit", "_target"];
+
+//if (_unit != acre_player) exitWith {};
```
Comment (jonpas) [inline addons/sys_gsa/fnc_connectChildrenActions.sqf]: Remove?

File: addons/sys_gsa/fnc_externalAntennaPFH.sqf [hunk 8]
```diff
@@ -0,0 +1,43 @@
+#include "script_component.hpp"
+/*
+ * Author: ACRE2Team
+ * Eventhandler to monitor if external radios can be still used by a specific player.
```
Comment (jonpas) [inline addons/sys_gsa/fnc_externalAntennaPFH.sqf]:
```suggestion
 * Event handler to monitor if external radios can be still used by a specific player.
```

File: addons/sys_gsa/fnc_getConnectedGsa.sqf [hunk 9]
```diff
@@ -0,0 +1,27 @@
+#include "script_component.hpp"
+/*
+ * Author: ACRE2Team
+ * Returns the connected gsa antenna object
```
Comment (jonpas) [inline addons/sys_gsa/fnc_getConnectedGsa.sqf]:
```suggestion
 * Returns the connected ground spike antenna antenna object.
```

File: addons/sys_gsa/fnc_handleMast.sqf [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+ * Public: No
+ */
+
+params ["_player", "_gsa", "_mountMast", ["_connectedRadio", ""]];
+
+if (_connectedRadio isEqualTo "") then {
+    _connectedRadio = _gsa getVariable [QGVAR(connectedRadio), ""];
+};
+
+// Temporarily disconnect the GSA from the radio.
```
Comment (jonpas) [inline addons/sys_gsa/fnc_handleMast.sqf]:
```suggestion
// Temporarily disconnect the GSA from the radio
```

File: addons/sys_gsa/fnc_hasItemMast.sqf [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+
+params ["_unit"];
+
+private _hasItemMast = false;
+
+{
+    if (_x isEqualTo "acre2_vhf30108mast") exitWith {
+        _hasItemMast = true;
+    };
+} forEach (items _unit);
```
Comment (jonpas) [inline addons/sys_gsa/fnc_hasItemMast.sqf]: Same story as above with unique items.

File: addons/sys_gsa/fnc_initGsa.sqf [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+ *
+ * Example:
+ * [cursorTarget] call acre_sys_gsa_fnc_initGsa
+ *
+ * Public: No
+ */
+
+params ["_gsa"];
+
+systemChat format ["gsa %1", _gsa];
```
Comment (jonpas) [inline addons/sys_gsa/fnc_initGsa.sqf]: Remove?

File: addons/sys_gsa/fnc_initGsa.sqf [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+if (!hasInterface) exitWith {};
+if (!isClass (configFile >> "CfgPatches" >> "ace_interaction")) exitWith {}; // No ACE exit.
```
Comment (jonpas) [inline addons/sys_gsa/fnc_initGsa.sqf]:
```suggestion
if (!isClass (configFile >> "CfgPatches" >> "ace_interaction")) exitWith {}; // No ACE exit
```

File: addons/sys_gsa/fnc_initRadio.sqf [hunk 14]
```diff
@@ -0,0 +1,20 @@
+#include "script_component.hpp"
+/*
+ * Author: ACRE2Team
+ * Initialises a radio with the additional state: externalAntennaConnected
```
Comment (jonpas) [inline addons/sys_gsa/fnc_initRadio.sqf]:
```suggestion
 * Initialises a radio with the additional state: externalAntennaConnected.
```

File: addons/sys_gsa/fnc_isAntennaConnected.sqf [hunk 15]
```diff
@@ -0,0 +1,21 @@
+#include "script_component.hpp"
+/*
+ * Author: ACRE2Team
+ * Checks if the ground spike antenna is connected to a radio.
+ *
+ * Arguments:
+ * 0: Unit <OBJECT> (Unused)
+ *^1: Ground spike antenna <OBJECT>
```
Comment (jonpas) [inline addons/sys_gsa/fnc_isAntennaConnected.sqf]:
```suggestion
 * 1: Ground spike antenna <OBJECT>
```

File: addons/sys_gsa/fnc_pickUp.sqf [hunk 16]
```diff
@@ -0,0 +1,58 @@
...
+ *
+ * Public: No
+ */
+
+params ["_unit", "_gsa"];
+
+private _radioId = _gsa getVariable [QGVAR(connectedRadio), ""];
+
+if (_radioId !=  "") then {
+    ["_unit", "_gsa"] call FUNC(disconnect);
```
Comment (jonpas) [inline addons/sys_gsa/fnc_pickUp.sqf]: Going to assume this is wrong.
```suggestion
    _this call FUNC(disconnect);
```

**Review by jonpas — CHANGES_REQUESTED** [review]
I am assuming compatibility with more antenna types will be added later, as in selection of antenna to place (or related to item you have in inventory)?

File: addons/sys_gsa/CfgVehicles.hpp [hunk 17]
```diff
@@ -0,0 +1,183 @@
...
+                    exceptions[] = {};
+                    priority = 5;
+                    //icon = "\idi\acre\addons\ace_interact\data\icons\antenna.paa";
+                };
+
+                class ACRE_mountMast {
+                    selection = "";
+                    displayName = CSTRING(mountMast);
+                    distance = 10;
+                    condition = QUOTE([ARR_2(_player), "ACRE_VHF30108MAST"] call DEFUNC(sys_core,hasItem));
```
Comment (jonpas) [inline addons/sys_gsa/CfgVehicles.hpp]: Usage of ARR doesn't look correct (`)` in wrong place) and double quotes inside `QUOTE`.

I suggest changing to:
```suggestion
                    condition = '[_player, "ACRE_VHF30108MAST"] call DEFUNC(sys_core,hasItem)';
```

File: addons/sys_gsa/CfgVehicles.hpp [hunk 18]
```diff
@@ -0,0 +1,183 @@
...
+                    statement = QUOTE([ARR_3(_player,_target, true)] call DFUNC(handleMast));
```
Comment (jonpas) [inline addons/sys_gsa/CfgVehicles.hpp]:
```suggestion
                    statement = QUOTE([ARR_3(_player,_target,true)] call DFUNC(handleMast));
```

**Review by jonpas — CHANGES_REQUESTED** [review]

File: addons/sys_antenna/CfgAcreComponents.hpp [hunk 19]
```diff
@@ -102,5 +113,24 @@ class CfgAcreComponents {
         connector = ACRE_CONNECTOR_3_8;
         height = 3.39;
         binaryGainFile = QPATHTOF(binary\AS-1729_VRC_gain.aba);
+        compatibleRadios[] = {"ACRE_VRC64"};
+    };
+
+    class ACRE_243CM_VHF_TNC: ACRE_BaseAntenna {
+        name = "2.43 Meter VHF Ground Spike Antenna";
```
Comment (jonpas) [inline addons/sys_antenna/CfgAcreComponents.hpp:120]: Stringtable all those antennas?

File: addons/sys_components/fnc_findAntenna.sqf [hunk 20]
```diff
@@ -39,15 +39,20 @@ private _searchFunction = {
                         if (_componentType == ACRE_COMPONENT_ANTENNA) then {
                             private _componentObject = _componentParentId;
                             if (IS_STRING(_componentParentId)) then {
-                                if (HASH_HASKEY(_attributes, "worldObject")) then {
-                                    _componentObject = HASH_GET(_attributes, "worldObject");
+                                private _groundSpikeAntenna = [_componentParentId] call EFUNC(sys_gsa,getConnectedGsa);
+                                if (isNull _groundSpikeAntenna) then {
+                                    if (HASH_HASKEY(_attributes, "worldObject")) then {
```
Comment (jonpas) [inline addons/sys_components/fnc_findAntenna.sqf]:
```suggestion
                                    if (HASH_HASKEY(_attributes,"worldObject")) then {
```
Macros don't like spaces.

File: addons/sys_components/fnc_findAntenna.sqf [hunk 21]
```diff
@@ -39,15 +39,20 @@ private _searchFunction = {
...
+                                        _componentObject = HASH_GET(_attributes, "worldObject");
```
Comment (jonpas) [inline addons/sys_components/fnc_findAntenna.sqf]:
```suggestion
                                        _componentObject = HASH_GET(_attributes,"worldObject");
```
Macros don't like spaces.

File: addons/sys_core/fnc_hasItem.sqf [hunk 22]
```diff
@@ -0,0 +1,25 @@
...
+ * Public: No
+ */
+#include "script_component.hpp"
+
+params [["_unit", objNull], ["_item", ""]];
+
+if (!GVAR(aceLoaded)) then {
+    _item in ([_unit] call FUNC(uniqueItems))
+} else {
+    _item in ([_unit] call ace_common_fnc_uniqueItems);
```
Comment (jonpas) [inline addons/sys_core/fnc_hasItem.sqf]:
```suggestion
    _item in ([_unit] call ace_common_fnc_uniqueItems)
```

File: addons/sys_gsa/data/models/acre_antennabag.rvmat (hunk 6 above)
Comment (jonpas) [reply addons/sys_gsa/data/models/acre_antennabag.rvmat]: Bump.

File: addons/sys_gsa/XEH_postInit.sqf [hunk 23]
```diff
@@ -0,0 +1,75 @@
+#include "script_component.hpp"
+
+[QGVAR(disconnectGsa), {
+    params ["_gsa", "_unit"];
```
Comment (jonpas) [inline addons/sys_gsa/XEH_postInit.sqf:4]: Functionality both?

File: addons/sys_gsa/fnc_getConnectedGsa.sqf [hunk 24]
```diff
@@ -0,0 +1,27 @@
+#include "script_component.hpp"
+/*
+ * Author: ACRE2Team
+ * Returns the connected ground spike antenna object
```
Comment (jonpas) [inline addons/sys_gsa/fnc_getConnectedGsa.sqf]:
```suggestion
 * Returns the connected ground spike antenna object.
```

**Review by jonpas — CHANGES_REQUESTED** [review]
Could use some documentation, in this or separate PR. Especially noting that ACE3 is required at this time.

File: addons/sys_antenna/CfgAcreComponents.hpp (hunk 19 above)
Comment (TheMagnetar) [reply addons/sys_antenna/CfgAcreComponents.hpp:120]: They are not even shown. It's only internal

File: addons/sys_gsa/XEH_postInit.sqf (hunk 23 above)
Comment (jonpas) [reply addons/sys_gsa/XEH_postInit.sqf:4]: Functions I meant, what the heck haha!

Comment (TheMagnetar) [reply addons/sys_gsa/XEH_postInit.sqf:4]: I was wondering...

**Review by jonpas — APPROVED** [review]

**Pull request merged**